### PR TITLE
is(): correct is() for types

### PR DESCRIPTION
**Pull request by filipsajdak**
This change fixes the inspection of variables.

* correct inspection of polymorphic types,
* correct inspection of pointers,
* add support for nullptr_t

### CODE REVIEW

File: include/cpp2util.h [hunk 1]
```diff
@@ -1629,39 +1629,44 @@ constexpr auto is( X&& ) {
...
 template< typename C, typename X >
-auto is( X const& x ) -> bool {
+constexpr auto is( X const& x ) -> auto {
     if constexpr (
         std::is_same_v<C, X>
         || std::is_base_of_v<C, X>
     )
     {
-        return true;
+        return std::true_type{};
```
Comment (hsutter) [inline include/cpp2util.h:1638]: What are the benefits of returning `std::true_type` instead of `bool`?

My familiarity with `std::true_type` and `std::false_type` is with TMP, and in general the compile-time work we're doing here tries to avoid TMP.

File: include/cpp2util.h [hunk 2]
```diff
@@ -1629,39 +1629,44 @@ constexpr auto is( X&& ) {
...
     }
     else if constexpr (
-        std::is_base_of_v<X, C>
-        || (
-            std::is_polymorphic_v<C>
-            && std::is_polymorphic_v<X>
-            )
+        std::is_polymorphic_v<C>
+        && std::is_polymorphic_v<X>
```
Comment (hsutter) [inline include/cpp2util.h:1642]: Is the `is_base_of` alternative redundant? Or is this fixing the semantics of a certain use case?

File: include/cpp2util.h [hunk 3]
```diff
@@ -1629,39 +1629,44 @@ constexpr auto is( X&& ) {
...
+        && std::is_pointer_v<X>
+    )
+    {
+        if (x != nullptr) {
+            return bool{is<std::remove_pointer_t<C>>(*x)};
+        }
         return false;
     }
+    else {
+        return std::false_type{};
```
Comment (hsutter) [inline include/cpp2util.h:1668]: As above: Can you summarize the benefit of `false_type` over `bool`?

Also, I haven't fully checked, just asking: Are the choices of `[true|false]_type` vs `bool` paired so that the former or the latter is consistently used for a given kind of successful/failed test?

File: include/cpp2util.h [hunk 4]
```diff
@@ -1629,39 +1629,44 @@ constexpr auto is( X&& ) {
...
+        && std::is_polymorphic_v<X>
     )
     {
-        if constexpr (std::is_pointer_v<X>) {
-            return Dynamic_cast<C const*>(x) != nullptr;
-        }
-        else {
-            return Dynamic_cast<C const*>(&x) != nullptr;
-        }
+        return Dynamic_cast<C const*>(&x) != nullptr;
```
Comment (hsutter) [inline include/cpp2util.h:1645]: Ah, because `is_pointer_v<X>` has to be `false` in this branch so it's dead code?

File: include/cpp2util.h (hunk 3 above)
Comment (gregmarr) [reply include/cpp2util.h:1668]: I believe that `true_type/false_type` means that the result is known at compile time because it's based solely on the types involved, and that `true/false` means that it's deferred to runtime because it depends on the value of the variable.

File: include/cpp2util.h (hunk 1 above)
Comment (gregmarr) [reply include/cpp2util.h:1638]: I commented on the same question below.

File: include/cpp2util.h (hunk 3 above)
Comment (filipsajdak) [reply include/cpp2util.h:1668]: Using `false_type`/`true_type` over `bool` gives the possibility of using the result in the compile-time. So, the result can be used in the `if constexpr`s or `requires` clause.

That means we know the result of some `is()` calls at compile time, which is useful for inspecting the types.

File: include/cpp2util.h (hunk 4 above)
Comment (filipsajdak) [reply include/cpp2util.h:1645]: In the following code:
```
        if constexpr (std::is_pointer_v<X>) {
             return Dynamic_cast<C const*>(x) != nullptr;
         }
```
the dead code is `return Dynamic_cast<C const*>(x) != nullptr;`

As `std::is_pointer_v<X>` will always be `false` just because to get to this `if`, we need to fulfill the following expression:
```
           (
            std::is_polymorphic_v<C>
            && std::is_polymorphic_v<X>
             )
```
That means `X` is polymorphic, and that means it is not a pointer (as a pointer does not fulfill the `is_polymorphic_v` type traits (see: https://godbolt.org/z/3j73cfWYe).

File: include/cpp2util.h (hunk 1 above)
Comment (filipsajdak) [reply include/cpp2util.h:1638]: Answered here: https://github.com/hsutter/cppfront/pull/1204#discussion_r1712342613

File: include/cpp2util.h (hunk 2 above)
Comment (filipsajdak) [reply include/cpp2util.h:1642]: The issue with `std::is_base_of_v<X, C>` is that it will also match non-polymorphic types for which dynamic_cast is undefined.

see: https://godbolt.org/z/GMvfrGzEh

Comment (filipsajdak) [reply include/cpp2util.h:1642]: Ah, the following check
```
std::is_polymorphic_v<C>
        && std::is_polymorphic_v<X>
```
is enough to know that we can use `dynamic_cast` and it works for all combinations of polymorphic types.

See: https://godbolt.org/z/fhTbE31Ke

Comment (filipsajdak) [reply include/cpp2util.h:1642]: I want to let you know that this PR is just the beginning of a series of PRs that correct or improve the `is()` and `as()`. The upcoming PRs (some of which I already requested) will improve/correct some of the cases that are still in this code.

I intend to make these changes digestive.

**Comment by hsutter** [comment]
Thanks! I think all our review questions are resolved -- are there any more commits I should wait for before merging, or can I go ahead and merge now?

**Comment by filipsajdak** [comment]
@hsutter please merge. I will rebase other PRs on top of this PR. thank you!

**Comment by hsutter** [comment]
Thanks!
